cli/lib/kontena/cli/services/exec_command.rb

```diff
@@ -69,16 +47,98 @@ def execute
         end
         return ret
       elsif instance
-        if !(container = service_containers.find{|container| container['instance_number'] == instance})
+        if !(container = service_containers.find{|c| c['instance_number'] == instance})
           exit_with_error "Service #{name} does not have container instance #{instance}"
         elsif container['status'] != 'running'
           exit_with_error "Service #{name} container #{container['name']} is not running, it is #{container['status']}"
-        else
+        elsif interactive?
+          interactive_exec(container)
+        else 
           exec_container(container)
         end
       else
-        exec_container(running_containers.first)
+        if interactive?
+          interactive_exec(running_containers.first)
+        else 
+          exec_container(running_containers.first)
+        end
+      end
+    end
+
+    # Exits if exec returns with non-zero
+    # @param [Hash] container
+    def exec_container(container)
+      exit_status = nil
+      if !silent? && (verbose? || all?)
+        spinner "Executing command on #{container['name']}" do
+          exit_status = normal_exec(container)
+
+          raise Kontena::Cli::SpinAbort if exit_status != 0
+        end
+      else
+        exit_status = normal_exec(container)
+      end
+
+      exit exit_status if exit_status != 0 && !skip?
+
+      return exit_status == 0
+    end
+
+    # @param [Hash] container
+    # @return [Boolean]
+    def normal_exec(container)
+      base = self
+      cmd = JSON.dump({ cmd: cmd_list })
+      exit_status = nil
+      token = require_token
+      url = ws_url(container['id'])
+      url << '?shell=true' if shell?
+      ws = connect(url, token)
```

Comment:
Missing `ws.on :error` handler to abort the CLI.

If any of the `on :open` etc handlers raises an error, then the CLI just dies silently, because the `Websocket::Client::Simple` thread dies, and so the main CLI thread just gets stuck waiting for it.